lib/state.js

```diff
@@ -157,21 +162,34 @@ StateResolver.prototype.capture_ = function() {
                                     result.error, true)
         };
       } else {
-        evaluated = that.resolveVariable_(expression, result.mirror);
+        evaluated = that.resolveVariable_(expression, result.mirror, true);
+        var varTableIdx = evaluated.varTableIndex;
+        if (typeof varTableIdx !== 'undefined'){
+          evalIndexSet.add(varTableIdx);
+        }
       }
       that.evaluatedExpressions_[index] = evaluated;
     });
   }
 
+  // The frames are resolved after the evaluated expressions so that 
+  // evaluated expressions can be evaluated as much as possible within 
+  // the max data size limits
+  var frames = that.resolveFrames_();
+
   // Now resolve the variables
   var index = MESSAGE_TABLE.length; // skip the sentinel values
   var noLimit = that.config_.capture.maxDataSize === 0;
-  while (index < that.rawVariableTable_.length && // NOTE: length changes in loop
-         (that.totalSize_ < that.config_.capture.maxDataSize || noLimit)) {
+  while (index < that.rawVariableTable_.length) { // NOTE: length changes in loop
     assert(!that.resolvedVariableTable_[index]); // shouldn't have it resolved yet
+    var isEvaluated = evalIndexSet.has(index);
     that.resolvedVariableTable_[index] =
-      that.resolveMirror_(that.rawVariableTable_[index]);
+      that.resolveMirror_(that.rawVariableTable_[index], isEvaluated);
     index++;
+
+    if (!noLimit && that.totalSize_ >= that.config_.capture.maxDataSize) {
```

Comment:
nit: readability would be improved if you reversed `noLimit` to `haveLimit` above and avoided a `!` here.